### PR TITLE
[3.x] Add support for saving multiple Images in BakedLightmap

### LINKED ISSUE #51741: BakedLightmap atlas size not respected
### Godot version

3.4.beta.3 mono

### System information

Windows 10

### Issue description

BakedLightmap node is not respecting the atlas size as you can see: 
BakedLightmapErrorDimensions
Also i remenber in the older version there was an error message that tells you that the size of the uv's were too big, in thet version i haven't see that error anymore.

### Steps to reproduce

Bake a scene with 3d models imported with low lightmap texel size (for example 0.01) and light baking as "Gen Lightmaps". Then bake

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by JFonS**
Instead of fitting all atlas slices into a single image, which meant there was a hard limit on the size, `BakedLightmap` will now save as many images as needed to fit all the slices generated by the lightmapper.

I only did some quick testing on a single scene, so it would be good to get this tested on more scenarios before merging.

The size limit is much less of an issue in `master` since the maximum size for images is much larger there. Still, once we see everything is working fine in `3.x`, this can be forward-ported to `master`.

Closes #51741, implements [godot-proposals#2147](https://github.com/godotengine/godot-proposals/issues/2147).

### CODE REVIEW

**Comment by Calinou** [comment]
I tried this on Reflection, but the resulting bake looks broken:



[test_lightmap_large_3.x.zip](https://github.com/godotengine/godot/files/8064047/test_lightmap_large_3.x.zip)

**Comment by JFonS** [comment]
I force-pushed a fix for @Calinou's issue, everything should be working now.

File: thirdparty/stb_rect_pack/stb_rect_pack.h [hunk 1]
```diff
@@ -575,7 +575,10 @@ STBRP_DEF int stbrp_pack_rects(stbrp_context *context, stbrp_rect *rects, int nu
 
    // set was_packed flags and all_rects_packed status
    for (i=0; i < num_rects; ++i) {
-      rects[i].was_packed = !(rects[i].x == STBRP__MAXVAL && rects[i].y == STBRP__MAXVAL);
+      // -- GODOT start --
+      // rects[i].was_packed = !(rects[i].x == STBRP__MAXVAL && rects[i].y == STBRP__MAXVAL);
+      rects[i].was_packed = !(rects[i].x == (int) STBRP__MAXVAL && rects[i].y == (int) STBRP__MAXVAL);
```
Comment (akien-mga) [inline thirdparty/stb_rect_pack/stb_rect_pack.h:580]: Could you PR that fix upstream too?

File: core/math/geometry.cpp [hunk 2]
```diff
@@ -36,6 +36,7 @@
 #include "thirdparty/misc/clipper.hpp"
 #include "thirdparty/misc/triangulator.h"
 #define STB_RECT_PACK_IMPLEMENTATION
+#define STBRP_LARGE_RECTS
```
Comment (akien-mga) [inline core/math/geometry.cpp:39]: Worth noting that stb_rect_pack 1.01 would make this the default behavior (we have 1.00): https://github.com/nothings/stb/blob/af1a5bc352164740c1cc1354942b1c6b72eacb8a/stb_rect_pack.h#L44

**Comment by JFonS** [comment]
@Akien `stb_rect_pack` 1.01 doesn't have the warning issue because it gets rid of the dual modes. 

Since updating to 1.01 fixes both changes I had made in this PR, I just removed them and we can do the upgrade in a separate PR.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by KiwwiPity** [comment]
There are any advantages a part from reducing memory usage? (if understood correctly)

**Comment by Calinou** [comment]
> Looks awesome! There are any advantages a part from reducing memory usage? (if understood correctly)

Supporting multiple atlases makes it possible to successfully bake lightmaps in large levels without exceeding the maximum texture size. Before multiple atlases were supported, you had to disable atlases to successfully bake lightmaps in a large scene. This was less efficient and resulted in a lot of files being created when baking lightmaps.

**Comment by gdlq** [comment]
Doesn't work in 3.5 RC2. 
Same error as before **Failed determining lightmap size. Maximum lightmap size too small?**
I tried the test scene from https://github.com/godotengine/godot/files/8064047/test_lightmap_large_3.x.zip
Added 10 additional cubes from blender to the scene and changed import settings in the Lightmap Texel Size to 0.05, I should have several exr textures but I got error Maximum lightmap size too small, if I change Lightmap Texel Size to 1.0 then it bakes all in single exr texture.

**Comment by Calinou** [comment]
@gdlq Please open a new issue with a minimal reproduction project attached.

**Pull request merged**